File: src/control/policy/RecoveryPolicyResolver.ts
import type { ServiceSnapshot } from '../../node/data/ServiceSnapshot.js'
import type { RecoveryDecision, ServiceRecoveryPolicy } from './ServiceRecoveryPolicy.js'

export class RecoveryPolicyResolver {
  public resolve(snapshot: ServiceSnapshot, policy: ServiceRecoveryPolicy): RecoveryDecision {
    if (snapshot.lifecycleState === policy.expectedState && snapshot.healthy) {
      return 'healthy'
    }
    if (policy.restartAllowed && policy.maxRestartAttempts > 0 && (snapshot.lifecycleState === 'stopped' || snapshot.lifecycleState === 'failed')) {
      return 'restart'
    }
    if (snapshot.lifecycleState === 'unknown') {
      return 'human'
    }
    return 'investigate'
  }
}

File: src/main/java/org/tavall/recovery/policy/RecoveryPolicyResolver.java
package org.tavall.recovery.policy;

import org.tavall.recovery.node.RecoveryNodeSnapshot;

import java.util.Locale;
import java.util.Objects;

/** Pure policy resolver. It never performs node mutations or consumes restart budget. */
public final class RecoveryPolicyResolver {
    public RecoveryDecision resolve(
            RecoveryNodeSnapshot.RecoveryServiceSnapshot snapshot,
            ServiceRecoveryPolicy policy
    ) {
        Objects.requireNonNull(snapshot, "snapshot");
        Objects.requireNonNull(policy, "policy");
        verifyTarget(snapshot, policy);

        String lifecycleState = normalized(snapshot.lifecycleState());
        if (lifecycleState.equals(normalized(policy.expectedState())) && snapshot.healthy()) {
            return RecoveryDecision.HEALTHY;
        }
        if (policy.restartAllowed()
                && policy.maxRestartAttempts() > 0
                && ("stopped".equals(lifecycleState) || "failed".equals(lifecycleState))) {
            return RecoveryDecision.RESTART;
        }
        if ("unknown".equals(lifecycleState)) {
            return RecoveryDecision.HUMAN_REQUIRED;
        }
        return RecoveryDecision.INVESTIGATE;
    }

    public boolean isHealthy(
            RecoveryNodeSnapshot.RecoveryServiceSnapshot snapshot,
            ServiceRecoveryPolicy policy
    ) {
        return resolve(snapshot, policy) == RecoveryDecision.HEALTHY;
    }

    private static void verifyTarget(
            RecoveryNodeSnapshot.RecoveryServiceSnapshot snapshot,
            ServiceRecoveryPolicy policy
    ) {
        if (!policy.nodeId().equals(snapshot.nodeId()) || !policy.serviceId().equals(snapshot.serviceId())) {
            throw new IllegalArgumentException(
                    "Recovery observation target " + snapshot.nodeId() + "/" + snapshot.serviceId()
                            + " does not match policy " + policy.nodeId() + "/" + policy.serviceId()
            );
        }
    }

    private static String normalized(String value) {
        return value == null ? "unknown" : value.trim().toLowerCase(Locale.ROOT);
    }
}
